control/control.go

```diff
@@ -889,40 +962,95 @@ func (p *pluginControl) sendPluginSubscriptionEvent(taskID string, pl core.Plugi
 		return serror.New(err)
 	}
 	e := &control_event.PluginSubscriptionEvent{
-		TaskId:           taskID,
-		PluginType:       int(pt),
-		PluginName:       pl.Name(),
-		PluginVersion:    pl.Version(),
-		SubscriptionType: int(strategy.UnboundSubscriptionType),
-	}
-	if pl.Version() > 0 {
-		e.SubscriptionType = int(strategy.BoundSubscriptionType)
+		TaskId:        taskID,
+		PluginType:    int(pt),
+		PluginName:    pl.Name(),
+		PluginVersion: pl.Version(),
 	}
+
 	if _, err := p.eventManager.Emit(e); err != nil {
 		return serror.New(err)
 	}
 	return nil
 }
 
+// UnsubscribeDeps calculates diff between previous state in task view (plugins and metrics) and current state.
+// Based on results, task view is updated and no longer valid plugins are unsubscribed
 func (p *pluginControl) UnsubscribeDeps(taskID string, mts []core.Metric, plugins []core.Plugin) []serror.SnapError {
 	var serrs []serror.SnapError
-	// If no metrics to unsubscribe then skip this section. Avoids errors when
-	// workflow is distributed and each node may not have metrics.
-	if len(mts) > 0 {
-		collectors, errs := p.gatherCollectors(mts)
-		if len(errs) > 0 {
-			serrs = append(serrs, errs...)
+	// calculate difference between previous state and current state
+	diffPlgs := diffPlugins(p.taskView[taskID].plugins, plugins)
+	// update view with current plugins and metrics
+	serr := p.updateViewOnUnsubs(taskID, mts, []core.RequestedMetric{}, plugins)
+	if serr != nil {
+		serrs = append(serrs, serr)
+	}
+
+	// unsubscribe only plugins which are not longer valid
+	for _, sub := range diffPlgs {
+		pool, err := p.pluginRunner.AvailablePlugins().getPool(fmt.Sprintf("%s:%s:%d", sub.TypeName(), sub.Name(), sub.Version()))
+		if err != nil {
+			serrs = append(serrs, err)
+			return serrs
 		}
-		for _, gc := range collectors {
-			plugins = append(plugins, gc.plugin)
+		if pool != nil {
+			pool.Unsubscribe(taskID)
+		}
+		serr := p.sendPluginUnsubscriptionEvent(taskID, sub)
+		if serr != nil {
+			serrs = append(serrs, serr)
 		}
 	}
 
-	for _, sub := range plugins {
+	// / this is the case for task stop (unsubscribe called from scheduler.stopTask)
```

Comment:
Why does it matter where unsubscribe is called from? Is there different behavior based on where it is called from? If so does it make sense for it to be the same function?